File: src/main/java/io/github/kingg22/api/vacunas/panama/web/controller/PdfController.java
package io.github.kingg22.api.vacunas.panama.web.controller;

import io.github.kingg22.api.vacunas.panama.response.ApiResponseCode;
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory;
import io.github.kingg22.api.vacunas.panama.response.ApiResponseUtil;
import io.github.kingg22.api.vacunas.panama.response.IApiResponse;
import io.github.kingg22.api.vacunas.panama.service.IPacienteService;
import io.github.kingg22.api.vacunas.panama.service.IVacunaService;
import io.github.kingg22.api.vacunas.panama.service.PdfService;
import io.github.kingg22.api.vacunas.panama.web.dto.DosisDto;
import io.github.kingg22.api.vacunas.panama.web.dto.PacienteDto;
import java.io.IOException;
import java.io.Serializable;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.context.request.ServletWebRequest;

@Slf4j
@RestController
@RequiredArgsConstructor
@RequestMapping(path = "/vacunacion/v1/pdf")
public class PdfController {
    private final PdfService pdfService;
    private final IPacienteService pacienteService;
    private final IVacunaService vacunaService;
    private final ApiResponseFactory apiResponseFactory;

    @GetMapping
    public ResponseEntity<byte[]> getPdfFile(
            @RequestParam("idPaciente") UUID idPaciente, @RequestParam("idVacuna") UUID idVacuna) {
        try {
            List<DosisDto> dosisDtos = vacunaService.getDosisByIdPacienteIdVacuna(idPaciente, idVacuna);
            if (dosisDtos.isEmpty()) {
                log.debug(dosisDtos.toString());
                return ResponseEntity.badRequest().build();
            }
            PacienteDto pacienteDto = pacienteService.getPacienteDtoById(idPaciente);
            UUID idCertificado = UUID.randomUUID();
            byte[] pdfStream = pdfService.generatePdf(pacienteDto, dosisDtos, idCertificado);

            HttpHeaders headers = new HttpHeaders();
            headers.add(
                    HttpHeaders.CONTENT_DISPOSITION,
                    "attachment; filename=certificado_vacunas_"
                            .concat(idCertificado.toString())
                            .concat(".pdf"));
            headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_PDF_VALUE);

            return ResponseEntity.ok().headers(headers).body(pdfStream);
        } catch (RuntimeException | IOException e) {
            log.debug(e.getMessage(), e);
            return ResponseEntity.internalServerError().build();
        }
    }

    @GetMapping("/base64")
    public ResponseEntity<IApiResponse<String, Serializable>> getPdfBase64(
            @RequestParam("idPaciente") UUID idPaciente,
            @RequestParam("idVacuna") UUID idVacuna,
            ServletWebRequest webRequest) {
        IApiResponse<String, Serializable> apiResponse = apiResponseFactory.createResponse();
        try {
            List<DosisDto> dosisDtos = vacunaService.getDosisByIdPacienteIdVacuna(idPaciente, idVacuna);
            if (dosisDtos.isEmpty()) {
                log.debug(dosisDtos.toString());
                apiResponse.addError(
                        ApiResponseCode.NOT_FOUND,
                        "Dosis de la vacuna para el paciente no fueron encontradas para generar el PDF");
                apiResponse.addStatusCode(HttpStatus.NOT_FOUND);
            }
            PacienteDto pDetalle = pacienteService.getPacienteDtoById(idPaciente);
            UUID idCertificado = UUID.randomUUID();
            String pdfBase64 = pdfService.generatePdfBase64(pDetalle, dosisDtos, idCertificado);
            apiResponse.addData("idCertificado", idCertificado.toString());
            apiResponse.addData("pdf", pdfBase64);
            apiResponse.addStatusCode(HttpStatus.OK);
            return ApiResponseUtil.sendResponse(apiResponse, webRequest);
        } catch (RuntimeException | IOException e) {
            log.debug(e.getMessage(), e);
            apiResponse.addError(HttpStatus.INTERNAL_SERVER_ERROR.toString(), "Ha ocurrido un error al generar el PDF");
            apiResponse.addStatusCode(HttpStatus.INTERNAL_SERVER_ERROR);
            return ApiResponseUtil.sendResponse(apiResponse, webRequest);
        }
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/modules/pdf/controller/PdfController.kt
package io.github.kingg22.api.vacunas.panama.modules.pdf.controller

import io.github.kingg22.api.vacunas.panama.modules.paciente.service.PacienteService
import io.github.kingg22.api.vacunas.panama.modules.pdf.service.PdfService
import io.github.kingg22.api.vacunas.panama.modules.vacuna.dto.DosisDto
import io.github.kingg22.api.vacunas.panama.modules.vacuna.service.VacunaService
import io.github.kingg22.api.vacunas.panama.response.ApiResponseCode
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory.createApiErrorBuilder
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory.createResponse
import io.github.kingg22.api.vacunas.panama.response.ApiResponseUtil.createResponseEntity
import io.github.kingg22.api.vacunas.panama.util.logger
import io.vertx.ext.web.RoutingContext
import jakarta.annotation.security.RolesAllowed
import jakarta.enterprise.context.RequestScoped
import jakarta.inject.Inject
import jakarta.ws.rs.GET
import jakarta.ws.rs.Path
import jakarta.ws.rs.QueryParam
import jakarta.ws.rs.core.Response
import org.eclipse.microprofile.jwt.JsonWebToken
import java.util.UUID

@Path("/pdf")
@RequestScoped
@RolesAllowed(value = ["PACIENTE_READ"])
class PdfController(
    private val pdfService: PdfService,
    private val pacienteService: PacienteService,
    private val vacunaService: VacunaService,
) {
    private val log = logger()

    @Inject
    lateinit var jwt: JsonWebToken

    @GET
    suspend fun getPdfFile(@QueryParam("idVacuna") idVacuna: UUID): Response {
        try {
            val (dosisDtos, idPaciente) = getDosisVacunaOfPaciente(idVacuna)

            if (dosisDtos.isEmpty()) {
                log.debug(dosisDtos.toString())
                return Response.status(404).build()
            }

            val pacienteDto = pacienteService.getPacienteDtoById(idPaciente)

            if (pacienteDto == null) {
                return Response.status(404).build()
            }

            val idCertificado: UUID = UUID.randomUUID()
            val pdfStream = pdfService.generatePdf(pacienteDto, dosisDtos, idCertificado)

            return Response.ok(pdfStream)
                .type("application/pdf")
                .header("Content-Disposition", "attachment; filename=\"certificado_vacunas_$idCertificado.pdf\"")
                .build()
        } catch (e: Exception) {
            log.error(e.message, e)
            return Response.status(500).build()
        }
    }

    @Path("/base64")
    @GET
    suspend fun getPdfBase64(@QueryParam("idVacuna") idVacuna: UUID, rc: RoutingContext): Response {
        val apiResponse = createResponse()
        try {
            val (dosisDtos, idPaciente) = getDosisVacunaOfPaciente(idVacuna)

            if (dosisDtos.isEmpty()) {
                log.debug(dosisDtos.toString())
                apiResponse.addError(
                    createApiErrorBuilder {
                        withCode(ApiResponseCode.NOT_FOUND)
                        message = "Dosis de la vacuna para el paciente no fueron encontradas para generar el PDF"
                    },
                )
                apiResponse.addStatusCode(404)
                return createResponseEntity(apiResponse, rc)
            }

            val pDetalle = pacienteService.getPacienteDtoById(idPaciente)

            if (pDetalle == null) {
                apiResponse.addError(
                    createApiErrorBuilder {
                        withCode(ApiResponseCode.NOT_FOUND)
                        message = "El paciente no fue encontrado, intente nuevamente."
                    },
                )
                return createResponseEntity(apiResponse, rc)
            }

            val idCertificado: UUID = UUID.randomUUID()
            val pdfBase64 = pdfService.generatePdfBase64(pDetalle, dosisDtos, idCertificado)

            apiResponse.addData("id_certificado", idCertificado.toString())
            apiResponse.addData("pdf", pdfBase64)
            apiResponse.addStatusCode(200)
        } catch (e: Exception) {
            log.debug(e.message, e)
            apiResponse.addError(
                createApiErrorBuilder {
                    withCode(ApiResponseCode.INTERNAL_SERVER_ERROR)
                    message = "Ha ocurrido un error al generar el PDF"
                },
            )
            apiResponse.addStatusCode(500)
        }
        return createResponseEntity(apiResponse, rc)
    }

    private suspend fun getDosisVacunaOfPaciente(idVacuna: UUID): Pair<List<DosisDto>, UUID> {
        val personaIdString = jwt.getClaim<String>("persona")
        check(personaIdString != null) { "Persona ID is null in JWT claims with ID: ${jwt.tokenID}" }

        val idPaciente: UUID = UUID.fromString(personaIdString)
        return vacunaService.getDosisByIdPacienteIdVacuna(idPaciente, idVacuna) to idPaciente
    }
}
